### ISSUE TITLE
make the swarm implementation configurable.

### ISSUE THREAD

**Issue by wenchaopeng** [libp2p/go-libp2p#875]
Hi：
may I ask when can swarm support configurable？

```
// TODO: Make the swarm implementation configurable.
swrm := swarm.NewSwarm(ctx, pid, cfg.Peerstore, cfg.Reporter)
if cfg.Filters != nil {
	swrm.Filters = cfg.Filters
}
return swrm, nil
```

**Comment by Stebalien** [comment]
This is not planned in the near future as we're not sure that the current swarm/host split is even correct. If you need a custom swarm, I recommend you construct a `Host` manually, passing in a custom swarm that way.

**Comment by wenchaopeng** [comment]
ok,thanks for your reply
I want to build a private network and refuse the connection of nodes that are not in the whitelist. I understand that there is a `libp2p.PrivateNetwork (prot pnet.Protector)` in the configuration, but `pnet.Protector` can only use IP + PORT(the parameter of this function is net.conn) . as I want use peerid as whilelist, do you have any good suggestions？

//The following code is my current implementation
```
type Protector struct {
	whiteList []string
}

func NewProtector() *Protector {
	return &Protector{}
}

func (pt *Protector) Protect(s net.Conn) (net.Conn, error) {
	remoteAddr := s.RemoteAddr().String()
	for _, v := range pt.whiteList {
		if v == remoteAddr {
			return s, nil
		}
	}
	return nil, fmt.Errorf("remote addr %s is not in whiteList", remoteAddr)
}

func (pt *Protector) Fingerprint() []byte {
	return nil
}
```

**Comment by Stebalien** [comment]
Note: `pnet.Protector` actually uses a shared key, not the IP+port. To filter by IP+port, you'd use the `Filter`s. Well, I _guess_ you could use it to filter by IP, but that's not what it's there for.

It sounds like you're looking for https://github.com/libp2p/go-libp2p/issues/872 which is being actively worked on by @aarshkshah1992. That will allow you to reject peers based on arbitrary requirements (IP address, peer ID, etc.).

**Comment by wenchaopeng** [comment]
yes,#872 is what I am looking for,I'll follow it,thanks very much
